### PR TITLE
[BC] Standardization of Transforms/Functionals

### PR DESCRIPTION
**Pull request by jamarshon**
## Removal of Legacy Transforms
-Compose: From https://github.com/pytorch/audio/issues/110, "we wouldn't need it because once things are based on Layers people can simply build a nn.Sequential()."
-Aliases and deprecated modules: SPECTROGRAM, F2M, MEL If we are going to start breaking things in this new release, then it makes sense to do as much as possible so that the future stable versions do not have technical debt
-BLC2CBL: there is no input/output to this library that uses BLC or CBL format (going to keep LC2CL mainly because all the inputs are (c, n) meaning it would be convenient for users to switch this format).

## Channel x Time
The input (some outputs) of all transforms/functions would be (c, n) e.g Spectrogram, MFCC, MelSpectrogram, Resample, etc. This is to be consistent with PyTorch which has channel followed by the number of samples.
This means removing the channel dimension flags that some transforms were using which could accept either (n, c) or (c, n).

## Channel x Frequency x Time
The output of stft is (c, f, t, 2) meaning for each channel, the columns are the fft of a certain window so as we travel horizontally we can see each column (fft) change over time. This matches the output of librosa so we no longer need to transpose in our test comparisions with spectrogram, melscale, and mfcc. Previously it was time x frequency.
image

## DCT and Filterbank matrices
Going to keep them as they are. These matrices should be shaped so that they can be multiplied without a transpose e.g for filterbank (..., n_freqs) x (n_freqs, n_mels) meaning each column is a filterbank and DCT (..., n_mels) x (n_mels, n_mfcc) meaning each column is a cos function

## Misc
Going to keep scale, padtrim, downmixmono as while they are simple, they could be operations that are performed often so it would be convenient to have the code to reduce amount of code the user has to write as well as provides documentation as the name of the transforms/functions communicates the intent clearly

### CODE REVIEW

**Comment by cpuhrsch** [comment]
@keunwoochoi @ksanjeevan please let us know what you think. This is going to resolve a lot of discussion :)

**Comment by ksanjeevan** [comment]
@cpuhrsch on the dimension side all great 👍. There still will be much to discuss on some variable/transform names I guess.

**Comment by keunwoochoi** [comment]
Great! All the dimension configs are aligned with what we've discussed in the `contrib` repo :) 

If the var names were an issue, we reached at [this conclusion](https://github.com/keunwoochoi/torchaudio-contrib/issues/36#issuecomment-502144987) if you wanna borrow some idea @cpuhrsch @jamarshon. They're probably slightly more verbose than usual, and that's what we wanted so that even beginners in either Pytorch or audio processing would get it only by reading the code.

File: test/test_transforms.py [hunk 1]
```diff
@@ -127,43 +98,37 @@ def test_mu_law_companding(self):
         sig_exp = transforms.MuLawExpanding(quantization_channels)(sig_mu)
         self.assertTrue(sig_exp.min() >= -1. and sig_exp.max() <= 1.)
 
-        repr_test = transforms.MuLawEncoding(quantization_channels)
```
Comment (vincentqb) [inline test/test_transforms.py]: Was this block meant to be removed?

Comment (jamarshon) [reply test/test_transforms.py]: Yes the check was to see if the class has a __repr__. Every class has a __repr__ so it did not  really do anything

**Comment by jamarshon** [comment]
In terms of variable name standardization, some concepts were taken from https://github.com/keunwoochoi/torchaudio-contrib/issues/36#issuecomment-502144987, mainly:

`waveform`: a tensor of audio samples
`sample_rate`: the rate of audio samples (samples per second)
`specgram`: a tensor of spectrogram
`mel_specgram`: a mel spectrogram
`hop_length`: the number of samples between the starts of consecutive frames
`n_freqs`: the number of bins in a linear spectrogram (i.e., n_freqs == n_fft // 2 + 1)
`min_freq`: the lowest frequency of the lowest band in a mel/CQT spectrogram
`max_freq`: the highest frequency of the highest band in a mel/CQT spectrogram

Other variable names
`n_fft`: the number of fourier bins (matches torch.stft naming)
`win_length`: the length of the STFT window (matches torch.stft naming)
`n_mfcc`, `n_mels`: just to be consistent with other n_* variables

Currently there is no complex tensors anywhere so no `complex_specgram`. Also no batching so no `waveforms`, `specgrams`, etc.

**Comment by jamarshon** [comment]
@vincentqb @keunwoochoi could you take a look and let me know what you think?

**Comment by keunwoochoi** [comment]
Looks good to me. The things mentioned in [the original thread](https://github.com/pytorch/audio/pull/152#issue-298981045) and [the comment](https://github.com/pytorch/audio/pull/152#issuecomment-513968536) seem aligned to what we had discussed :) 
Curious though - what do you mean by no batching?

**Comment by cpuhrsch** [comment]
@keunwoochoi - see [this comment](https://github.com/pytorch/audio/pull/131#discussion_r305991797) for a wider discussion around this topic.

**Comment by jamarshon** [comment]
https://github.com/pytorch/audio/pull/131#discussion_r305991797
None of the current transforms/functional receive batched inputs/outputs. I think the goal is to have no batch support in this version and then add batching in the next version

**Comment by vincentqb** [comment]
I'm happy with the convention, and the code looks good to me. Thanks for updating the code and the tests.

File: torchaudio/functional.py [hunk 2]
```diff
@@ -9,26 +9,23 @@
...
+    'create_fb_matrix',
     'create_dct',
-    'BLC2CBL',
     'mu_law_encoding',
     'mu_law_expanding'
 ]
 
 
 @torch.jit.script
 def scale(tensor, factor):
```
Comment (cpuhrsch) [inline torchaudio/functional.py]: Do we even still want to keep this one? I mean, it's just "tensor / factor" with an "to(torch.float32)" before.

File: torchaudio/functional.py [hunk 3]
```diff
@@ -40,52 +37,44 @@ def scale(tensor, factor):
...
-        ch_dim (int): Dimension of channel (not size)
+        waveform (torch.Tensor): Tensor of audio of size (c, n)
 
     Returns:
-        torch.Tensor: Mono signal
+        torch.Tensor: Tensor that has been downmixed of size (1, n)
     """
-    if not tensor.is_floating_point():
-        tensor = tensor.to(torch.float32)
+    if not waveform.is_floating_point():
```
Comment (cpuhrsch) [inline torchaudio/functional.py]: I'm not a big fan of these implicit conversions. What if the users isn't aware this needs floats? Also, it'll return floats for an integer input. Didn't we agree on all waveforms to be floats in range -1 to 1?

File: torchaudio/functional.py [hunk 4]
```diff
@@ -254,59 +244,93 @@ def istft(stft_matrix,          # type: Tensor
...
+    if normalized:
         spec_f /= window.pow(2).sum().sqrt()
-    spec_f = spec_f.pow(power).sum(-1)  # get power of "complex" tensor (c, l, n_fft)
+    spec_f = spec_f.pow(power).sum(-1)  # get power of "complex" tensor
     return spec_f
 
 
 @torch.jit.script
-def create_fb_matrix(n_stft, f_min, f_max, n_mels):
+def spectrogram_to_DB(specgram, multiplier, amin, db_multiplier, top_db=None):
```
Comment (cpuhrsch) [inline torchaudio/functional.py]: Did you just move this around? Why?

File: torchaudio/transforms.py [hunk 5]
```diff
@@ -7,441 +7,362 @@
...
 
-    def __repr__(self):
-        return self.__class__.__name__ + '()'
+        Returns:
+            torch.Tensor: Tensor that has been downmixed of size (1, n)
+        """
+        return F.downmix_mono(waveform)
 
 
 class LC2CL(torch.jit.ScriptModule):
```
Comment (cpuhrsch) [inline torchaudio/transforms.py]: What about this: If we agree that all signal must be of form C x L and torch.load produces C x L then any user should be responsible for this conversion herself? - Similar to scale, this wraps a simple function 'transpose'.

File: torchaudio/transforms.py [hunk 6]
```diff
@@ -7,441 +7,362 @@
...
-        Args:
-        sr (int) : sample rate of audio signal
-        n_mfcc (int) : number of mfc coefficients to retain
-        dct_type (int) : type of DCT (discrete cosine transform) to use
-        norm (string, optional) : norm to use
-        log_mels (bool) : whether to use log-mel spectrograms instead of db-scaled
-        melkwargs (dict, optional): arguments for MelSpectrogram
-    """
-    __constants__ = ['sr', 'n_mfcc', 'dct_type', 'top_db', 'log_mels']
+class MelScale(torch.jit.ScriptModule):
```
Comment (cpuhrsch) [inline torchaudio/transforms.py]: Can you move this higher up again? It'll be much easier to review the changes.

File: torchaudio/functional.py (hunk 3 above)
Comment (jamarshon) [reply torchaudio/functional.py]: Done.

File: torchaudio/functional.py (hunk 2 above)
Comment (jamarshon) [reply torchaudio/functional.py]: Done.

File: torchaudio/transforms.py (hunk 5 above)
Comment (jamarshon) [reply torchaudio/transforms.py]: Done.

File: torchaudio/functional.py (hunk 4 above)
Comment (jamarshon) [reply torchaudio/functional.py]: https://github.com/pytorch/audio/blob/master/torchaudio/functional.py#L11
It makes sense to have `spectrogram` and `spectrogram_to_DB` beside each other and `create_fb_matrix` and `create_dct` beside each other.

**Comment by jamarshon** [comment]
Removing scale and lc2cl because we want to assume the user will do the necessary formatting of their input. Also want to keep library lean and only have complex, multiline algorithms.

Removing downmixmono from feedback. Related PR https://github.com/keunwoochoi/torchaudio-contrib/pull/45

File: torchaudio/functional.py (hunk 4 above)
Comment (cpuhrsch) [reply torchaudio/functional.py]: ok, then lets do that in a follow-up PR, but not while changing them at the same time

**Comment by vincentqb** [comment]
If we are making the code leaner, should we only have DBScale/MelScale transforms without having the SpectrogramDB/MelSpectrogram if they are easy to chain?

**Comment by vincentqb** [comment]
> If we are making the code leaner, should we only have DBScale/MelScale transforms without having the SpectrogramDB/MelSpectrogram if they are easy to chain?

If we do this, we'll do it as part of a separate PR.

**Comment by vincentqb** [comment]
Here are things we can consider for a later PR:

- [ ] Remove SpectrogramToDB, and create a DBScale transformation.
- [ ] Remove MelSpectrogram and keep only MelScale.
- [ ] Remove PadTrim.
- [ ] Introduce Padding-only transformation.

File: test/test_functional.py [hunk 7]
```diff
@@ -197,54 +195,6 @@ def _num_stft_bins(signal_len, fft_len, hop_length, pad):
...
-    (torch.randn(1, 100000)),
-    (torch.randn(1, 2, 100000)),
-    pytest.param(torch.randn(1, 100), marks=xfail(raises=RuntimeError)),
-])
-@pytest.mark.parametrize('pad_mode', [
-    # 'constant',
-    'reflect',
-])
-@unittest.skipIf(not IMPORT_LIBROSA, 'Librosa is not available')
-def test_stft(waveform, fft_length, hop_length, pad_mode):
```
Comment (cpuhrsch) [inline test/test_functional.py]: Why might want to bring this back / add it to core later on for the regular torch.stft

**Comment by jamarshon** [comment]
> Based on the work done here, here are things we can consider for a later PR:
> 
> * [ ]  Remove SpectrogramToDB, and create a DBScale transformation [like so](https://github.com/keunwoochoi/torchaudio-contrib/blob/master/torchaudio_contrib/functional.py#L277).
> * [ ]  Remove MelSpectrogram and keep only MelScale.
> * [ ]  Remove PadTrim.
> * [ ]  Introduce Padding-only transformation, [like so](https://github.com/pytorch/audio/pull/122/files#diff-a3428ec41790ed12faea36fcc6dd8f06R457).
> * [ ]  Rename MuLawExpanding to MuLawDecoding.

Rename MuLawExpanding to MuLawDecoding. https://github.com/pytorch/audio/pull/159

**Comment by keunwoochoi** [comment]
awesome!!!

**Pull request merged**